aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -222,91 +213,94 @@ Tensor& comparison_op_out(Tensor& result, const Tensor& self, const Tensor& othe
       check_convert(self.item(), other.scalar_type());
     }
   }
-  return native::comparison_op_impl_out(result, self, other, stub);
+  auto iter = TensorIterator::comparison_op(result, self, other, /*check_mem_overlap=*/true);
+  stub(iter.device_type(), iter);
+  return result;
 }
 
-template <typename Stub>
-Tensor comparison_op(const Tensor& self, const Tensor& other, Stub& stub) {
+template <typename OutImpl>
+Tensor comparison_op(const Tensor& self, const Tensor& other, OutImpl& out_impl) {
   Tensor result = at::empty({0}, self.options().dtype(kBool));
-  return native::comparison_op_out(result, self, other, stub);
+  return out_impl(result, self, other);
 }
 
 // To avoid overflow during type promotion we will check that both dtypes of self and other are same
-template <typename Stub>
-Tensor& comparison_op_(Tensor& self, const Tensor& other, Stub& stub) {
+template <typename OutImpl>
+Tensor& comparison_op_(Tensor& self, const Tensor& other, OutImpl& out_impl) {
   TORCH_CHECK(self.dtype() == other.dtype(),
               "Expected object of scalar type ", self.dtype(), " but got scalar type ",
               other.dtype(), " for argument 'other'");
-  return native::comparison_op_impl_out(self, self, other, stub);
+  return out_impl(self, self, other);
 }
 
 // validates that is possible to convert Scalar other to self's dtype without overflow.
 // This behavior is unique to comparison ops; arithmetic operations don't do this.
 // In the future, we should reconsider this inconsistency and decide if we want to add the same check to arithmetic ops.
-template <typename Stub>
-Tensor& comparison_op_out(Tensor& result, const Tensor& self, Scalar other, Stub& stub) {
-  return native::comparison_op_out(result, self, wrapped_scalar_tensor_and_check_convert(other, self), stub);
+template <typename OutImpl>
+Tensor& comparison_op_out(Tensor& result, const Tensor& self, Scalar other, OutImpl& out_impl) {
+  return out_impl(result, self, wrapped_scalar_tensor_and_check_convert(other, self));
 }
 
-template <typename Stub>
-Tensor comparison_op(const Tensor& self, Scalar other, Stub& stub) {
-  Tensor result = at::empty({0}, self.options().dtype(kBool));
-  return native::comparison_op_out(result, self, other, stub);
+template <typename OutImpl>
+Tensor comparison_op(const Tensor& self, Scalar other, OutImpl& out_impl) {
+  return comparison_op(self, wrapped_scalar_tensor_and_check_convert(other, self), out_impl);
 }
 
-template <typename Stub>
-Tensor& comparison_op_(Tensor& self, Scalar other, Stub& stub) {
-  return native::comparison_op_impl_out(self, self, wrapped_scalar_tensor_and_check_convert(other, self), stub);
+template <typename OutImpl>
+Tensor& comparison_op_(Tensor& self, Scalar other, OutImpl& out_impl) {
+  return out_impl(self, self, wrapped_scalar_tensor_and_check_convert(other, self));
 }
 
+using OutFunc = std::add_const<Tensor&(&)(Tensor&, const Tensor&, const Tensor&)>::type;
```

Comment:
Without looking at this code closely, could you explain/add a comment why this static_cast is necessary? e.g. the unary one didn't have to do it. Thanks!

Reply:
Comment added